packages/tenant-process/src/services/tenantService.ts

```diff
@@ -0,0 +1,115 @@
+  toCreateEventTenantUpdated,
+} from "../model/domain/toEvent.js";
+import { ApiSelfcareTenantSeed } from "../model/types.js";
+import { selfcareIdConflict } from "../model/domain/errors.js";
+import {
+  assertResourceAllowed,
+  getTenantKind,
+  getTenantKindLoadingCertifiedAttributes,
+} from "./validators.js";
+import { ReadModelService } from "./readModelService.js";
+
+// eslint-disable-next-line @typescript-eslint/explicit-function-return-type
+export function tenantServiceBuilder(
+  config: TenantProcessConfig,
+  readModelService: ReadModelService
+) {
+  const repository = eventRepository(
+    initDB({
+      username: config.eventStoreDbUsername,
+      password: config.eventStoreDbPassword,
+      host: config.eventStoreDbHost,
+      port: config.eventStoreDbPort,
+      database: config.eventStoreDbName,
+      schema: config.eventStoreDbSchema,
+      useSSL: config.eventStoreDbUseSSL,
+    }),
+    tenantEventToBinaryData
+  );
+
+  return {
+    async selfcareUpsertTenant({
+      tenantSeed,
+      authData,
+    }: {
+      tenantSeed: ApiSelfcareTenantSeed;
+      authData: AuthData;
+    }): Promise<string> {
+      const existingTenant = await readModelService.getTenantByExternalId(
+        tenantSeed.externalId
+      );
+      if (existingTenant) {
+        await assertResourceAllowed(existingTenant.data.id, authData);
+        const tenantKind = await getTenantKindLoadingCertifiedAttributes(
+          readModelService,
+          existingTenant.data.attributes,
+          existingTenant.data.externalId
+        );
+
+        const updatedSelfcareId = evaluateNewSelfcareId({
+          tenant: existingTenant.data,
+          newSelfcareId: tenantSeed.selfcareId,
+        });
+
+        const updatedTenant: Tenant = {
+          ...existingTenant.data,
+          kind: tenantKind,
+          selfcareId: updatedSelfcareId,
+          updatedAt: new Date(),
+        };
+
+        return await repository.createEvent(
+          toCreateEventTenantUpdated(
+            existingTenant.data.id,
+            existingTenant.metadata.version,
+            updatedTenant
+          )
+        );
+      } else {
+        const newTenant: Tenant = {
+          id: uuidv4(),
+          name: tenantSeed.name,
+          attributes: [],
+          externalId: tenantSeed.externalId,
+          features: [],
+          mails: [],
+          selfcareId: tenantSeed.selfcareId,
+          kind: getTenantKind([], tenantSeed.externalId),
+          createdAt: new Date(),
+        };
+        return await repository.createEvent(
+          toCreateEventTenantAdded(newTenant)
+        );
+      }
+    },
+  };
+}
+
+export type TenantService = ReturnType<typeof tenantServiceBuilder>;
+
+function evaluateNewSelfcareId({
+  tenant,
+  newSelfcareId,
+}: {
+  tenant: Tenant;
+  newSelfcareId: string;
+}): string {
+  if (!tenant.selfcareId) {
+    return newSelfcareId;
+  } else {
+    if (tenant.selfcareId !== newSelfcareId) {
```

Comment:
```suggestion
  if (tenant.selfcareId && tenant.selfcareId !== newSelfcareId) {
```

I would also move this functions in `validators`

Reply:
Updated